app/apollo/resolvers/group.js

```diff
@@ -413,35 +447,53 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, groupUuids, clusterIds }, `${queryName} validating`);
-
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
-        groupUuids.forEach( value => validateString( 'groupUuids', value ) );
+        groupUuids.forEach( value => { validateString( 'groupUuids', value ); } );
         clusterIds.forEach( value => validateString( 'clusterIds', value ) );
 
-        // Create output for graphQL plugins
-        const groups = await models.Group.find({org_id, uuid: {$in: groupUuids}});
-        if (groups.length < 1) {
-          throw new NotFoundError(context.req.t('None of the passed group uuids were found'));
+        let groups = await getAllowedResources(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, null, groupUuids);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} retrieved allowed resources`);
+
+        if (groups.length != groupUuids.length) {
+          throw new NotFoundError(context.req.t('One or more of the passed group uuids were not found'));
+        }
+
+        if (groups.length < 1) { throw new NotFoundError(context.req.t('None of the passed group uuids were found')); }
+
+        let allAllowedClusters = false;
+        try {
+          await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context);
+          allAllowedClusters = true;
+        }
+        catch(e){ // If exception thrown, user does NOT have auth to all resources of this type, and code must later filter based on fine grained auth
+        }
+        logger.info({req_id, user, org_id, groupUuids, clusterIds, allAllowedClusters}, `${queryName} validating - allAllowed: ${allAllowedClusters}`);
+
+        let clusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
+
+        if (!allAllowedClusters){
+          clusters = await filterResourcesToAllowed(me, org_id, ACTIONS.ATTACH, TYPES.CLUSTER, clusters, context);
+          logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} filtered resources to allowed`);
         }
+
+        // Create output for graphQL plugins
         const groupObjs = _.map(groups, (group)=>{
           return {
             name: group.name,
             uuid: group.uuid,
           };
         });
-        const clusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
-        const clusterObjs = _.map(clusters, (cluster)=>{
+        clusters = _.map(clusters, (cluster)=>{
```

Comment:
Instead of redefining `clusters`, should keep the original definition of a new variable `const clusterObjs = _.map(clusters, (cluster)=....`.  This is makes the code more intelligible, as the `clusters` object doesn't represent different types of objects at different places in the code.  It also keeps the code in sync with the earlier `assignClusterGroups` implementation, which also defines `clusterObjs` variable for use with `pluginContext`.

Reply:
changed to clusterObjs

Reply:
resolved